<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en-us" lang="en-us">
<head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  
  <title>TCGA RNA Expression &middot; GRIP</title>
  

  
  <link rel="stylesheet" href="https://docs.bmeg.io/grip/css/poole.css">
  <link rel="stylesheet" href="https://docs.bmeg.io/grip/css/darcula.css">
  <link rel="stylesheet" href="https://docs.bmeg.io/grip/css/syntax.css">
  <link rel="stylesheet" href="https://docs.bmeg.io/grip/css/theme.css">
  <link rel="stylesheet" href="https://docs.bmeg.io/grip/css/funnel.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Sans:400,400italic,700">

  
  <link rel="apple-touch-icon-precomposed" sizes="144x144" href="https://docs.bmeg.io/grip/apple-touch-icon-144-precomposed.png">
  <link rel="shortcut icon" href="https://docs.bmeg.io/grip/favicon.png">

  <script src="https://docs.bmeg.io/grip/js/highlight.pack.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>

</head>
<body>

  <div class="global-header">
    <div class="global-header-container">
      <div class="global-header-home">
        <a href="https://docs.bmeg.io/grip"><h1>GRIP</h1></a>
      </div>
      <ul class="global-header-nav">
          <li><a href="https://docs.bmeg.io/grip/download//">Download</a></li>
          <li><a href="https://docs.bmeg.io/grip/docs/">Docs</a></li>
          <li><a href="https://github.com/bmeg/grip">GitHub</a></li>
          <li><a href="https://gitter.im/bmeg/grip">Chat</a></li>
        </ul>
      <div class="global-header-ohsucb">
        <a href="https://www.ohsu.edu/compbio/"><h2>OHSU Comp Bio</h2></a>
      </div>
    </div>
  </div>


<div class="content section group">

  <div class="sidebar col span_3_of_12">
    <ul class="sidebar-nav">
      
      
        
        <li>
          
          <a href="https://docs.bmeg.io/grip/download/"
             class="sidebar-nav-item "
          >Download</a></li>
          
        
        <li>
          
          <a href="https://docs.bmeg.io/grip/docs/"
             class="sidebar-nav-item "
          >Overview</a></li>
          
        
        <li>
          
          <span class="intermediate">Setup your Database</span></li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/databases/elastic/"
                   class="sidebar-nav-item "
                >Elasticsearch</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/databases/kvstore/"
                   class="sidebar-nav-item "
                >Embedded KV Store</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/databases/mongo/"
                   class="sidebar-nav-item "
                >MongoDB</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/databases/sql/"
                   class="sidebar-nav-item "
                >SQL</a></li>
            </ul>
          </li>
          
          
        
        <li>
          
          <span class="intermediate">Query a Graph</span></li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/queries/getting_started/"
                   class="sidebar-nav-item "
                >Getting Started</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/queries/operations/"
                   class="sidebar-nav-item "
                >Operations</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/queries/jsonpath/"
                   class="sidebar-nav-item "
                >Referencing Fields</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/queries/graphql/"
                   class="sidebar-nav-item "
                >GraphQL</a></li>
            </ul>
          </li>
          
          
        
        <li>
          
          <span class="intermediate">Tutorials</span></li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/tutorials/amazon/"
                   class="sidebar-nav-item "
                >Amazon Purchase Network</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/tutorials/tcga-rna/"
                   class="sidebar-nav-item  active"
                >TCGA RNA Expression</a></li>
            </ul>
          </li>
          
          
        
        <li>
          
          <span class="intermediate">Security</span></li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://docs.bmeg.io/grip/docs/security/basic/"
                   class="sidebar-nav-item "
                >Basic Auth</a></li>
            </ul>
          </li>
          
          
        
        <li>
          
          <span class="intermediate">Developers</span></li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://github.com/bmeg/grip/blob/master/gripql/gripql.proto"
                   class="sidebar-nav-item "
                >API</a></li>
            </ul>
          </li>
          
          <li>
            <ul class="sidebar-nav sidebar-nav-nested">
              <li>
                <a href="https://godoc.org/github.com/bmeg/grip"
                   class="sidebar-nav-item "
                >GoDoc</a></li>
            </ul>
          </li>
          
          
        
      
    </ul>
  </div>

  <div class="main col span_8_of_12">
    

<h3 id="explore-tcga-rna-expression-data">Explore TCGA RNA Expression Data</h3>

<p>Create the graph</p>

<pre><code>grip create tcga-rna
</code></pre>

<p>Get the data</p>

<pre><code>curl -O http://download.cbioportal.org/gbm_tcga_pub2013.tar.gz
tar xvzf gbm_tcga_pub2013.tar.gz
</code></pre>

<p>Load clinical data</p>

<pre><code>./example/load_matrix.py tcga-rna gbm_tcga_pub2013/data_clinical.txt --row-label 'Donor'
</code></pre>

<p>Load RNASeq data</p>

<pre><code>./example/load_matrix.py tcga-rna gbm_tcga_pub2013/data_RNA_Seq_v2_expression_median.txt -t  --index-col 1 --row-label RNASeq --row-prefix &quot;RNA:&quot; --exclude RNA:Hugo_Symbol
</code></pre>

<p>Connect RNASeq data to Clinical data</p>

<pre><code>./example/load_matrix.py tcga-rna gbm_tcga_pub2013/data_RNA_Seq_v2_expression_median.txt -t  --index-col 1 --no-vertex --edge 'RNA:{_gid}' rna
</code></pre>

<p>Connect Clinical data to subtypes</p>

<pre><code>./example/load_matrix.py tcga-rna gbm_tcga_pub2013/data_clinical.txt --no-vertex -e &quot;{EXPRESSION_SUBTYPE}&quot; subtype --dst-vertex &quot;{EXPRESSION_SUBTYPE}&quot; Subtype
</code></pre>

<p>Load Hugo Symbol to EntrezID translation table from RNA matrix annotations</p>

<pre><code>./example/load_matrix.py tcga-rna gbm_tcga_pub2013/data_RNA_Seq_v2_expression_median.txt --column-include Entrez_Gene_Id --row-label Gene
</code></pre>

<p>Load Mutation Information</p>

<pre><code>./example/load_matrix.py tcga-rna gbm_tcga_pub2013/data_mutations_extended.txt --skiprows 1 --index-col -1  --regex Matched_Norm_Sample_Barcode '\-\d\d$' '' --edge '{Matched_Norm_Sample_Barcode}' variantIn --edge '{Hugo_Symbol}' effectsGene --column-exclude ma_func.impact ma_fi.score MA_FI.score MA_Func.Impact MA:link.MSA MA:FImpact MA:protein.change MA:link.var MA:FIS MA:link.PDB --row-label Variant
</code></pre>

<p>Load Proneural samples into a matrix</p>

<pre><code class="language-python">import pandas
import gripql

conn = gripql.Connection(&quot;http://localhost:8201&quot;)
O = conn.graph(&quot;tcga-rna&quot;)
genes = {}
for k, v in O.query().V().where(gripql.eq(&quot;_label&quot;, &quot;Gene&quot;)).render([&quot;_gid&quot;, &quot;Hugo_Symbol&quot;]):
    genes[k] = v
data = {}
for row in O.query().V(&quot;Proneural&quot;).in_().out(&quot;rna&quot;).render([&quot;_gid&quot;, &quot;_data&quot;]):
    data[row[0]] = row[1]
samples = pandas.DataFrame(data).rename(genes).transpose().fillna(0.0)
</code></pre>

<h1 id="matrix-load-project">Matrix Load project</h1>

<pre><code>usage: load_matrix.py [-h] [--sep SEP] [--server SERVER]
                      [--row-label ROW_LABEL] [--row-prefix ROW_PREFIX] [-t]
                      [--index-col INDEX_COL] [--connect]
                      [--col-label COL_LABEL] [--col-prefix COL_PREFIX]
                      [--edge-label EDGE_LABEL] [--edge-prop EDGE_PROP]
                      [--columns [COLUMNS [COLUMNS ...]]]
                      [--column-include COLUMN_INCLUDE] [--no-vertex]
                      [-e EDGE EDGE] [--dst-vertex DST_VERTEX DST_VERTEX]
                      [-x EXCLUDE] [-d]
                      db input

positional arguments:
  db                    Destination Graph
  input                 Input File

optional arguments:
  -h, --help            show this help message and exit
  --sep SEP             TSV delimiter
  --server SERVER       Server Address
  --row-label ROW_LABEL
                        Vertex Label used when loading rows
  --row-prefix ROW_PREFIX
                        Prefix added to row vertex gid
  -t, --transpose       Transpose matrix
  --index-col INDEX_COL
                        Column number to use as index (and gid for vertex
                        load)
  --connect             Switch to 'fully connected mode' and load matrix cell
                        values on edges between row and column names
  --col-label COL_LABEL
                        Column vertex label in 'connect' mode
  --col-prefix COL_PREFIX
                        Prefix added to col vertex gid in 'connect' mode
  --edge-label EDGE_LABEL
                        Edge label for edges in 'connect' mode
  --edge-prop EDGE_PROP
                        Property name for storing value when in 'connect' mode
  --columns [COLUMNS [COLUMNS ...]]
                        Rename columns in TSV
  --column-include COLUMN_INCLUDE
                        List subset of columns to use from TSV
  --no-vertex           Do not load row as vertex
  -e EDGE EDGE, --edge EDGE EDGE
                        Create an edge the connected the current row vertex
                        args: &lt;dst&gt; &lt;edgeType&gt;
  --dst-vertex DST_VERTEX DST_VERTEX
                        Create a destination vertex, args: &lt;dstVertex&gt;
                        &lt;vertexLabel&gt;
  -x EXCLUDE, --exclude EXCLUDE
                        Exclude row id
  -d                    Run in debug mode. Print actions and make no changes

</code></pre>

  </div>

</div>
</body>
</html>
